### ISSUE TITLE
error: context not equal to range's container

### ISSUE THREAD

**Issue by blahah** [apache/incubator-annotator#25]
Hi, I'm trying to use annotator to allow generating an annotation from any arbitrary selection in a web page.

I'm having an issue with `describeTextQuoteByRange` from `@annotator/dom`.

If I modify the demo so that the `context` passed to `describeTextQuoteByRange` is anything other than the original value (`selectable`), it fails with the error:

```
Uncaught (in promise) Error: Context not equal to range's container; not implemented.
    at quote.js:48
    at Generator.next (<anonymous>)
    at step (asyncToGenerator.js:10)
    at _next (asyncToGenerator.js:25)
    at asyncToGenerator.js:32
    at new Promise (<anonymous>)
    at asyncToGenerator.js:5
    at describeTextQuoteByRange (quote.js:21)
    at HTMLDocument.onSelectionChange (mount.js:26)
```

I copied the approach used in the demo - I create a range using `var selectableRange = document.createRange()`, then add a node using `selectableRange.selectNodeContents(el)`.

I tried using `document.body` as `el`, and various other elements larger than `selectable`, but they all give the same error. I also tried leaving out the `context` key from the options object, same result.

Can you provide any insight into how this should be done?

**Comment by Treora** [comment]
As you may be aware, annotator's code is still very immature, and the most relevant part of the error message is the 'not implemented'. So in any case the solution should be to develop this function further, not to change the way you invoke it. I see I put a bit more info in the [source code](https://github.com/apache/incubator-annotator/blob/f7e7daab30949dccbf990105c08bfe887b1b6db9/packages/dom/src/text/quote.mjs#L38-L40):
```
  // FIXME We should get range index relative to context. Look at
  // dom-anchor-text-position? For now, we implement the easy case where the
  // ranges are within the same container.
```

Would be good to sit down some time and develop this to a more usable state.

**Comment by BigBlueHat** [comment]
I did some digging and it seems like the demo code limits which nodes can be used in a way the underlying library code does not:
https://github.com/apache/incubator-annotator/blob/master/demo/index.js#L57-L59

I'm not sure if this `isWithinNode()` method is solely for the demo's selection box, or if something like that is always needed when using the underlying library.

This sounds like a job for @tilgovi! 😉 I'm happy to hack into it more, but some basic guidance would help me not make a mess. 😁

**Comment by Treora** [comment]
> I'm not sure if this isWithinNode() method is solely for the demo's selection box, or if something like that is always needed when using the underlying library.

It is needed in the demo, because the (content of) the `#selectable` text box is regarded as the 'document' we are annotating (and the `#corpus` box next to it is the document the annotation is reapplied to). Passing a range that falls outside the document you are working in seems generally invalid. So the `isWithinNode` check is done because we pass the contents of `#selectable` as the `context` to `describeRange`.

The actual issue here is more or less the other way around: if the context/'document' is larger than a single text node, e.g. the whole `document.body`, things should of course work fine, but that (most usual) situation not yet implemented.

**Comment by tilgovi** [comment]
Now that the build is in place, I'm comfortable importing dom-anchor-* libraries. I'll do that ASAP and then ping back this issue.

**Comment by BigBlueHat** [comment]
I'm curious how our recent work might have effected the status of this issue.

@tilgovi let's track that importing task in a new issue.

**Comment by tilgovi** [comment]
Fixed in c7e98eeac3e07e45b14b94e2aec2add8f137aea9.
